dbt/macros/dist_to_nearest_geometry.sql

```diff
@@ -0,0 +1,88 @@
+-- Macro that takes a `source_model` containing geometries and joins it
+-- against `spatial.parcel` in order to generate the distance from each PIN
+-- to each geometry and year combination
+--
+-- The `source_conditional` allows for filtering on the source table
+-- using standard SQL, passed as a string
+{% macro dist_to_nearest_geometry(source_model, source_conditional) %}
+
+    with
+        -- Universe of all possible PINs. This ignores years since PINs don't
+        -- actually move very often, so unique xy coordinates are a good enough
+        -- proxy for PIN coordinates and year. We do this to limit the number
+        -- of parcels for which we need to perform spatial operations
+        distinct_pins as (
+            select distinct x_3435, y_3435 from {{ source("spatial", "parcel") }}
+        ),
+
+        -- Years that exist for parcel data. This determines the set of years
+        -- for which data will be filled forward in time i.e. if park locations
+        -- exist for 2020, and distinct_years goes up to 2023, then park data
+        -- from 2020 will be filled forward to 2023
+        distinct_years as (select distinct year from {{ source("spatial", "parcel") }}),
+
+        -- Crosswalk of the source data and distinct years, used to perform
+        -- the forward filling described above
+        fill_years as (
+            select dy.year as pin_year, max(df.year) as fill_year
+            from {{ source_model }} as df
+            cross join distinct_years as dy
+            where
+                dy.year >= df.year
+                {% if source_conditional is defined %}
+                    and {{ source_conditional }}
+                {% endif %}
```

Comment:
I'm not sure if there's a less clunky way to do this, but this does in fact work.